lib/connection/connection.js

```diff
@@ -528,97 +528,142 @@ function merge(options1, options2) {
-    self.emit('connect', self);
+    return callback(null, connection);
   });
 
   // Set the options for the connection
   connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
   connection.setTimeout(self.connectionTimeout);
   connection.setNoDelay(self.noDelay);
 
-  return connection;
+  // Add handlers for events
+  connection.once('error', err =>
+    callback({ err: err, type: 'error', family: options.family }, null)
+  );
+  connection.once('timeout', err => callback({ err: err, type: 'timeout' }, null));
+  connection.once('close', err => callback({ err: err, type: 'close' }, null));
+  connection.on('data', err => callback({ err: err, type: 'data' }, null));
+  return;
 }
 
-function makeUnsecureConnection(self, family) {
-  // Create new connection instance
-  let connection_options;
-  if (self.domainSocket) {
-    connection_options = { path: self.host };
-  } else {
-    connection_options = { port: self.port, host: self.host };
-    connection_options.family = family;
-  }
-
-  const connection = net.createConnection(connection_options);
-
-  // Set the options for the connection
-  connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
-  connection.setTimeout(self.connectionTimeout);
-  connection.setNoDelay(self.noDelay);
-
-  connection.once('connect', function() {
-    // Set socket timeout instead of connection timeout
-    connection.setTimeout(self.socketTimeout);
-    // Emit connect event
-    self.emit('connect', self);
-  });
+function fastFallback(self, _options, callback) {
+  const options = prepareConnectionOptions(self, _options);
+
+  let connection;
+  const connectionHandler = (err, _connection) => {
+    const _errorHandler = errorHandler(self);
+    const _timeoutHandler = timeoutHandler(self);
+    const _closeHandler = closeHandler(self);
+    const _dataHandler = dataHandler(self);
+
+    if (err) {
+      switch (err.type) {
+        case 'error':
+          if (err.family === 6) {
+            return function() {
+              if (self.logger.isDebug()) {
+                self.logger.debug(
+                  f(
+                    'connection %s for [%s:%s] errored out with [%s]',
+                    self.id,
+                    self.host,
+                    self.port,
+                    JSON.stringify(err)
+                  )
+                );
+              }
+            };
+          } else {
+            return _errorHandler(err.err);
+          }
+        case 'timeout':
+          return _timeoutHandler(err.err);
+        case 'close':
+          return _closeHandler(err.err);
+        case 'data':
+          return _dataHandler(err.err);
+      }
+    }
 
-  return connection;
-}
+    if (connection) {
+      _connection.removeAllListeners('error');
+      _connection.removeAllListeners('timeout');
+      _connection.removeAllListeners('close');
+      _connection.removeAllListeners('data');
+      _connection.end();
+      _connection.unref();
```

Comment:
you need to remember to `return` after the cleanup has happened, otherwise you will call the callback below with the selected connection twice